### ISSUE TITLE
Including PB.externalIncludePath in PB.protoSources leads to redundant class generation

### ISSUE THREAD

**Issue by adkafka** [akka/akka-grpc#628]
### Versions used 

Scala version: 2.12.8
Akka version: 2.5.23
Akka-gRPC version: 0.6.2

### Expected Behavior

When defining a `.proto` file that relies on a google type (`import "google/protobuf/wrappers.proto";`), the code generation should not regenerate these files, and it should use the compiled classes included in the runtime jar (`scalapb-runtime`).

### Actual Behavior

Due to [this line](https://github.com/akka/akka-grpc/blob/master/sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala#L118) (`PB.protoSources += PB.externalIncludePath.value`), all files in the `externalIncludePath` will be codegen'ed through ScalaPB. This results in regenerating the code for the google wrapper types, despite them being included in the runtime jar.

### Relevant logs
SBT logs that these external schemas are compiled:

```
[info] Compiling schema /app/target/protobuf_external/google/protobuf/compiler/plugin.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/descriptor.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/struct.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/any.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/duration.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/field_mask.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/source_context.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/type.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/empty.proto
[info] Compiling schema /app/target/protobuf_external/scalapb/scalapb.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/timestamp.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/wrappers.proto
[info] Compiling schema /app/target/protobuf_external/google/protobuf/api.proto
```

### Example
* Start with the [akka-grpc-quickstart-scala](https://developer.lightbend.com/guides/akka-grpc-quickstart-scala/) project
* Add an import to the `src/main/protobuf/helloworld.proto` file
  * `import "google/protobuf/wrappers.proto"`
* Change the `name` type in `HelloRequest` from `string` to `google.protobuf.StringValue`
* Add the following dependency to `build.sbt`
  * `"com.thesamet.scalapb"       %% "scalapb-runtime"    % scalapb.compiler.Version.scalapbVersion % "protobuf"`
* compile the project
* Notice the extra proto files generated in the logs, as well as in the target directory.

### Workaround

For now, we added the following block to our `build.sbt` files to get ScalaPB to ignore these proto files:

```
excludeFilter in PB.generate := new SimpleFileFilter((file: File) => {
  val excludeDirs = Set(
    "google/protobuf",
    "google/protobuf/compiler",
    "scalapb"
  )

  val parentDir = file.getParent
  excludeDirs.exists(p => parentDir.endsWith(p))
})
```

**Comment by adkafka** [comment]
@raboof Thanks for the reply. I'd be happy to submit a PR if we decide the workaround is sufficient. So to summarize:

1) We want akka-grpc to continue to generate classes for externally included files
2) We are willing to explicitly exclude `scalapb` and `google/protobuf` libraries from the list of classes

I'm fairly new to grpc, but I'm surprised that 1) is a common pattern. People would publish a jar of just `.proto` files? Seems like a more common pattern would be to publish the generated code in a jar. In your experience, is 1) common?

ScalaPB uses the `"protobuf"` configuration flag in `libraryDependencies` to flag a library as including some external protobuf files, but does not generate code for them. Maybe `akka-grpc` should use a separate configuration flag to indicate "external protobuf file that we need to generate code for" if we want to continue to support that pattern?

**Comment by raboof** [comment]
> I'm fairly new to grpc, but I'm surprised that 1) is a common pattern. People would publish a jar of just .proto files?

I think so, yes. The reason is there are many ways to convert a `.proto` file to code: of course Java code generation will produce different classes to Scala code generation, and there are parameters to the code generation (like `flat_package`) that influence what the generated classes look like. In your project, you may not like the choices the upstream library author made.

Are you sure your workaround works both when using `flat_package` and when not using that option? I'd kind of expect the in-project generated code to contain links to classes that don't exist in `scalapb-runtime` if that option doesn't match what was used to generate the classes for that library.

**Comment by bjaglin** [comment]
I also found this behavior strange and annoying as it has a strong impact on build times on multi-modules projects where **local, parent-project protos gets recompiled for each child module**.

I, for myself, reverted this behavior in a not-so-subtle manner via:
```
      .settings(
        // Avoid recompiling commons & external protos by cancelling AkkaGrpcPlugin's default of compiling dependencies
        // https://github.com/akka/akka-grpc/blob/0a53ed/sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala#L97-L98
        inConfig(Compile)(PB.protoSources := Seq(sourceDirectory.value / "protobuf")),
        inConfig(Test)(PB.protoSources := Seq(sourceDirectory.value / "protobuf")),
      )
```

> ScalaPB uses the "protobuf" configuration flag in libraryDependencies to flag a library as including some external protobuf files, but does not generate code for them. Maybe akka-grpc should use a separate configuration flag to indicate "external protobuf file that we need to generate code for" if we want to continue to support that pattern?

Agreed that overloading the semantics of `ScalaPB` is confusing. Another solution would be to discuss this upstream?

**Comment by raboof** [comment]
> multi-modules projects where local, parent-project protos gets recompiled for each child module

I agree in this context it makes more sense to generate the code only once. I wonder how this is done in Lagom projects.

> Agreed that overloading the semantics of `ScalaPB` is confusing. Another solution would be to discuss this upstream?

Might be interesting indeed!

**Comment by raboof** [comment]
I think there's use cases for both approaches, and indeed it might make sense to propose/contribute a more fine-grained feature for this upstream instead of dumping the external include path into the `protoSources` wholesale. I've submitted https://github.com/thesamet/sbt-protoc/issues/144 to get the discuss a way forward there.

**Comment by bjaglin** [comment]
I guess this can be closed as of 0.8.2

**Comment by raboof** [comment]
Yes :+1:
